sky/utils/command_runner.py

```diff
@@ -298,31 +298,34 @@ def rsync(
         # --filter
         rsync_command.append(RSYNC_FILTER_OPTION)
 
-        # --exclude-from
-        resolved_source = pathlib.Path(source).expanduser().resolve()
-        if (resolved_source / GIT_EXCLUDE).exists():
-            # Ensure file exists; otherwise, rsync will error out.
-            rsync_command.append(
-                RSYNC_EXCLUDE_OPTION.format(str(resolved_source / GIT_EXCLUDE)))
-
-        # rsync doesn't support '~' in a quoted target path. need to expand it.
-        full_source_str = str(resolved_source)
-        if resolved_source.is_dir():
-            full_source_str = os.path.join(full_source_str, '')
+        if up:
+            # The source is a local path, so we need to resolve it.
+            # --exclude-from
+            resolved_source = pathlib.Path(source).expanduser().resolve()
+            if (resolved_source / GIT_EXCLUDE).exists():
+                # Ensure file exists; otherwise, rsync will error out.
+                rsync_command.append(
+                    RSYNC_EXCLUDE_OPTION.format(
+                        str(resolved_source / GIT_EXCLUDE)))
 
         ssh_options = ' '.join(
             ssh_options_list(self.ssh_private_key, self.ssh_control_name))
         rsync_command.append(f'-e "ssh {ssh_options}"')
         # To support spaces in the path, we need to quote source and target.
         if up:
+            # rsync doesn't support '~' in a quoted target path. need to
+            # expand it.
+            full_source_str = str(resolved_source)
+            if resolved_source.is_dir():
+                full_source_str = os.path.join(full_source_str, '')
             rsync_command.extend([
                 f'{full_source_str!r}',
                 f'{self.ssh_user}@{self.ip}:{target!r}',
             ])
         else:
             rsync_command.extend([
-                f'{self.ssh_user}@{self.ip}:{full_source_str!r}',
-                f'{target!r}',
+                f'{self.ssh_user}@{self.ip}:{source!r}',
+                f'{os.path.expanduser(target)!r}',
```

Comment:
Just to confirm: when syncing down, `source` won't contain `~` right?

Reply:
Hmm, it can contain `~`, as we are downloading from `~/sky_logs/...`, but seems it work for me with the quote. The actual `rsync` command looks like this:
```
rsync -Pavz --filter='dir-merge,- .gitignore' -e "ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ConnectTimeout=30s -o ForwardAgent=yes -o ControlMaster=auto -o ControlPath=/tmp/skypilot_ssh_zhwu/d8bd79cc13/%C -o ControlPersist=300s" ubuntu@3.86.215.210:'~/sky_logs/sky-2022-12-12-12-51-20-038596/*' '/Users/zhwu/sky_logs/sky-2022-12-12-12-51-20-038596'
```

Do you know why it does not work with quote previously?

Reply:
Just tried again and looks like `~` won't be interpreted correctly with quotes by rsync when syncing from local to remote. 
```
(sky-tmp) weichiang@mbp sky % ls ~/repos/sky/nospacedir
a.txt

(sky-tmp) weichiang@mbp sky % rsync -Pavz --filter='dir-merge,- .gitignore' -e "ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ConnectTimeout=30s -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=300s"  '~/repos/sky/nospacedir' cpu:'~/abc'
Warning: Permanently added '44.212.16.43' (ED25519) to the list of known hosts.
building file list ...
rsync: link_stat "/Users/weichiang/repos/sky/~/repos/sky/nospacedir" failed: No such file or directory (2)
0 files to consider
```

However, syncing from remote to local seem to be fine.. maybe the behavior is just different.

Reply:
Great findings! Just added it to the comment. : )

Reply:
Great findings! Just added it to the comment. Thanks! : )